### PR TITLE
Use utf8mb4 MySQL character set

### PR DESCRIPTION
**Pull request by cedric-anne**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | -

Supports emoji, and, unless I am wrong, more Han ideographs.

image

Migration took only 15 minutes for a 1.5GB database on my developpement machine (which host DB on a HDD).

Remaining work TO DO:
 - [x] Check for `innodb_large_prefix` / `innodb_page_size` unsupported values.
 - [x] Force `ROW_FORMAT = Dynamic` in installation script / migration class.
 - [x] Check if `ROW_FORMAT` is `REDUNDANT` or `COMPACT` and block migration (or migrate to `ROW_FORMAT = Dynamic` if operation is not too risky).
 - [x] Do not use Litemoji lib when DB supports emojis.
 - [x] Test, test, and test again.

### CODE REVIEW

**Comment by tomolimo** [comment]
Hello @cedric-anne 
Thank you to have done this: it was also on my todo list, as we saw this new emoji are arriving more and more in our company...
Regards,
Tomolimo

**Comment by cedric-anne** [comment]
Migration took about 120 minutes for a 6.4GB database on my developpement machine (which host DB on a HDD):
 - 75 minutes for ROW_FORMAT=Compact to ROW_FORMAT=Dynamic,
 - 45 minutes for utf8mb4 tables conversion.

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -126,6 +126,16 @@ jobs:
...
+          - {php-version: "8.0", db-image: "mysql:5.6", always: true}
+          - {php-version: "8.0", db-image: "mysql:5.7", always: true}
+          - {php-version: "8.0", db-image: "mysql:8.0", always: true}
+          - {php-version: "8.0", db-image: "percona:5.6", always: true}
+          - {php-version: "8.0", db-image: "percona:5.7", always: true}
+          - {php-version: "8.0", db-image: "percona:8.0", always: true}
+          - {php-version: "8.0", db-image: "mariadb:10.1", always: true}
+          - {php-version: "8.0", db-image: "mariadb:10.2", always: true}
+          - {php-version: "8.0", db-image: "mariadb:10.3", always: true}
+          - {php-version: "8.0", db-image: "mariadb:10.4", always: true}
```
Comment (cedric-anne) [inline .github/workflows/ci.yml:138]: This should be removed before merge.
```suggestion
```

File: inc/dbconnection.class.php [hunk 2]
```diff
@@ -506,4 +576,69 @@ static function changeCronTaskStatus($enable = true) {
...
+      $charset = $use_utf8mb4 ? 'utf8mb4' : 'utf8';
+
+      $dbh->set_charset($charset);
+
+      // The mysqli::set_charset function will make COLLATE to be defined to the default one for used charset.
+      // As we are not using the default COLLATE, we have to define it using `SET NAMES` query.
+      switch ($charset) {
+         case 'utf8':
+            // Legacy charset, should be deprecated in next major version.
+            $dbh->query("SET NAMES 'utf8' COLLATE 'utf8_unicode_ci';");
```
Comment (trasher) [inline inc/dbconnection.class.php:600]: We could probably add a Toolbox::deprecated

Comment (cedric-anne) [reply inc/dbconnection.class.php:600]: Not now, because migration is not mandatory yet. Will have to be deprecated in the second next major version.

**Review by trasher — APPROVED** [review]

**Review by orthagh — CHANGES_REQUESTED** [review]
we maybe miss a warning on central inviting super-admin to do the migration.

**Review by orthagh — CHANGES_REQUESTED** [review]
i had a bunch of errors

```log
[2021-01-18 15:31:43] glpisqllog.ERROR: DBmysql::query() in /var/www/html/glpi/9.6.git/inc/dbmysql.class.php line 329
  *** MySQL query error:
  SQL: ALTER TABLE `glpi_notificationtemplates` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci
  Error: Incorrect datetime value: '0000-00-00 00:00:00' for column 'date_mod' at row 33
  Backtrace :
  inc/console/migration/utf8mb4command.class.php:267 
  inc/console/migration/utf8mb4command.class.php:86  Glpi\Console\Migration\Utf8mb4Command->migrateToUtf8mb4()
  vendor/symfony/console/Command/Command.php:255     Glpi\Console\Migration\Utf8mb4Command->execute()
  vendor/symfony/console/Application.php:1009        Symfony\Component\Console\Command\Command->run()
  inc/console/application.class.php:248              Symfony\Component\Console\Application->doRunCommand()
  vendor/symfony/console/Application.php:273         Glpi\Console\Application->doRunCommand()
  vendor/symfony/console/Application.php:149         Symfony\Component\Console\Application->doRun()
  bin/console:102                                    Symfony\Component\Console\Application->run()
  {"user":"@DESKTOP-CUCL3BV"} 
[2021-01-18 15:31:45] glpisqllog.ERROR: DBmysql::query() in /var/www/html/glpi/9.6.git/inc/dbmysql.class.php line 329
  *** MySQL query error:
  SQL: ALTER TABLE `glpi_plugin_eachange_requests` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci
  Error: Invalid default value for 'desired_end_date'
  Backtrace :
  inc/console/migration/utf8mb4command.class.php:267 
  inc/console/migration/utf8mb4command.class.php:86  Glpi\Console\Migration\Utf8mb4Command->migrateToUtf8mb4()
  vendor/symfony/console/Command/Command.php:255     Glpi\Console\Migration\Utf8mb4Command->execute()
  vendor/symfony/console/Application.php:1009        Symfony\Component\Console\Command\Command->run()
  inc/console/application.class.php:248              Symfony\Component\Console\Application->doRunCommand()
  vendor/symfony/console/Application.php:273         Glpi\Console\Application->doRunCommand()
  vendor/symfony/console/Application.php:149         Symfony\Component\Console\Application->doRun()
  bin/console:102                                    Symfony\Component\Console\Application->run()
  {"user":"@DESKTOP-CUCL3BV","mem_usage":"2.055\", 17.16Mio)"} 
[2021-01-18 15:31:45] glpisqllog.ERROR: DBmysql::query() in /var/www/html/glpi/9.6.git/inc/dbmysql.class.php line 329
  *** MySQL query error:
  SQL: ALTER TABLE `glpi_plugin_eachange_requests_tasks` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci
  Error: Invalid default value for 'planification_date'
  Backtrace :
  inc/console/migration/utf8mb4command.class.php:267 
  inc/console/migration/utf8mb4command.class.php:86  Glpi\Console\Migration\Utf8mb4Command->migrateToUtf8mb4()
  vendor/symfony/console/Command/Command.php:255     Glpi\Console\Migration\Utf8mb4Command->execute()
  vendor/symfony/console/Application.php:1009        Symfony\Component\Console\Command\Command->run()
  inc/console/application.class.php:248              Symfony\Component\Console\Application->doRunCommand()
  vendor/symfony/console/Application.php:273         Glpi\Console\Application->doRunCommand()
  vendor/symfony/console/Application.php:149         Symfony\Component\Console\Application->doRun()
  bin/console:102                                    Symfony\Component\Console\Application->run()
  {"user":"@DESKTOP-CUCL3BV","mem_usage":"0.004\", 17.16Mio)"} 
[2021-01-18 15:31:46] glpisqllog.ERROR: DBmysql::query() in /var/www/html/glpi/9.6.git/inc/dbmysql.class.php line 329
  *** MySQL query error:
  SQL: ALTER TABLE `glpi_plugin_formcreator_issues` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci
  Error: Invalid default value for 'date_mod'
  Backtrace :
  inc/console/migration/utf8mb4command.class.php:267 
  inc/console/migration/utf8mb4command.class.php:86  Glpi\Console\Migration\Utf8mb4Command->migrateToUtf8mb4()
  vendor/symfony/console/Command/Command.php:255     Glpi\Console\Migration\Utf8mb4Command->execute()
  vendor/symfony/console/Application.php:1009        Symfony\Component\Console\Command\Command->run()
  inc/console/application.class.php:248              Symfony\Component\Console\Application->doRunCommand()
  vendor/symfony/console/Application.php:273         Glpi\Console\Application->doRunCommand()
  vendor/symfony/console/Application.php:149         Symfony\Component\Console\Application->doRun()
  bin/console:102                                    Symfony\Component\Console\Application->run()
  {"user":"@DESKTOP-CUCL3BV","mem_usage":"1.151\", 17.16Mio)"} 
[2021-01-18 15:32:00] glpisqllog.ERROR: DBmysql::query() in /var/www/html/glpi/9.6.git/inc/dbmysql.class.php line 329
  *** MySQL query error:
  SQL: ALTER TABLE `glpi_tickets` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci
  Error: Incorrect datetime value: '0000-00-00 00:00:00' for column 'begin_waiting_date' at row 2917
  Backtrace :
  inc/console/migration/utf8mb4command.class.php:267 
  inc/console/migration/utf8mb4command.class.php:86  Glpi\Console\Migration\Utf8mb4Command->migrateToUtf8mb4()
  vendor/symfony/console/Command/Command.php:255     Glpi\Console\Migration\Utf8mb4Command->execute()
  vendor/symfony/console/Application.php:1009        Symfony\Component\Console\Command\Command->run()
  inc/console/application.class.php:248              Symfony\Component\Console\Application->doRunCommand()
  vendor/symfony/console/Application.php:273         Glpi\Console\Application->doRunCommand()
  vendor/symfony/console/Application.php:149         Symfony\Component\Console\Application->doRun()
  bin/console:102                                    Symfony\Component\Console\Application->run()
  {"user":"@DESKTOP-CUCL3BV","mem_usage":"13.591\", 17.16Mio)"} 
  ```

Note sure if it's a broken db on my side (it's very possible).

We may need to do a mini tutorial for plugin developpers also

**Comment by orthagh** [comment]
we saw last point together, issue is on my db only.

So just the central part to do and it's good to merge for me

**Comment by cedric-anne** [comment]
> ``  SQL: ALTER TABLE `glpi_notificationtemplates` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci
  Error: Incorrect datetime value: '0000-00-00 00:00:00' for column 'date_mod' at row 33``
>
> ``  SQL: ALTER TABLE `glpi_tickets` CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci
  Error: Incorrect datetime value: '0000-00-00 00:00:00' for column 'begin_waiting_date' at row 2917``

These errors are due to invalid data in DB, I do not even know how it is possible. This command only change collation, dans dates should not be affected. I think we cannot handle this kind of issues. Migration fails on this table, so admin has to manually fix the row and launch migration again.

Other errors are related to plugins db schema, and, again, how it is possible to have a running erronous db schema ? Why this error did not happened on table creation of on MySQL serveur update ?

> We may need to do a mini tutorial for plugin developpers also

Yes, I already added the "need_doc_improvement" flag to think about that. I will do it once PR will be merged.

**Comment by cedric-anne** [comment]
1) I added a warning on migration that explains that the migration havve to be executed separately. I had to change verbosity level of "warnings" to make them displayed on the default verbosity level.

2) I added a warning on central page, but only if "utf8mb4 mode" is already enabled, in order to warn admin when some tables (plugin's tables for instance) are not migrated.

3) I fixed the SQL query that compute tables to migrate. Indeed, when having multiple GLPI on same DB, it was joigning  `information_schema.columns` rows having `table_schema` different from `information_schema.tables` row.

**Review by trasher — APPROVED** [review]

**Review by orthagh — APPROVED** [review]

**Pull request merged**